relentless/optimize/method.py

```diff
@@ -206,81 +231,81 @@ def step_size(self, value):
         self._step_size = value
 
     @property
-    def max_iter(self):
-        """int: The maximum number of optimization iterations allowed."""
-        return self._max_iter
+    def line_search(self):
+        """:class:`LineSearch`: The line search object used to optimize the step size."""
+        return self._line_search
 
-    @max_iter.setter
-    def max_iter(self, value):
-        if not isinstance(value, int):
-            raise TypeError('The maximum number of iterations must be an integer.')
-        if value < 1:
-            raise ValueError('The maximum number of iterations must be positive.')
-        self._max_iter = value
+    @line_search.setter
+    def line_search(self, value):
+        if value is not None and not isinstance(value, LineSearch):
+            raise TypeError('If defined, the line search parameter must be a LineSearch object.')
+        self._line_search = value
 
+#TODO: document
 class LineSearch:
-    def __init__(self, abs_tol, max_step_size, max_iter):
+    """
+    """
+    def __init__(self, abs_tol, max_iter):
         self.abs_tol = abs_tol
-        self.max_step_size = max_step_size
         self.max_iter = max_iter
 
-    def find(self, objective, result_0=None):
-        dvars = objective.design_variables()
-        if result_0 is None:
-            result_0 = objective.compute()
-        dvars_0 = result_0.design_variables
+    def find(self, objective, res_start, res_end):
+        """
+        """
+        res_0 = objective.compute()
+        ovars = res_0.design_variables
 
         # compute normalized search direction, adjust variables based on max step size
-        d = {}
-        d_norm = np.linalg.norm(np.asarray(list(result_0._gradient.todict().values())))
-        for x in dvars:
-            d[x] = (-result_0.gradient(x))/d_norm
-            x.value += self.max_step_size*d[x]
-        result = objective.compute()
+        d = {x: res_end.design_variables[x]-res_start.design_variables[x] for x in ovars}
+        max_step = np.linalg.norm(np.asarray(list(d.values())))
+        for x in ovars:
+            d[x] /= max_step
+            x.value += max_step*d[x]
+        res = objective.compute()
 
         # compute initial and post-adjustment target values
         target_0 = 0
         target = 0
-        for x in dvars:
-            target_0 += (d[x]*-result_0.gradient(x))
-            target += (d[x]*-result.gradient(x))
+        for x in ovars:
+            target_0 += (d[x]*-res_0.gradient(x))
```

Comment:
FYI, if we want to use a relative tolerance, we would base it on this quantity, i.e. `np.abs(target) < rel_tol*np.abs(target_0)`